lib/internal/wrap_js_stream.js

```diff
@@ -0,0 +1,220 @@
+ * "real" JavaScript stream is needed. JSStreamWrap is exactly this.
+ */
+class JSStreamWrap extends Socket {
+  constructor(stream) {
+    const handle = new JSStream();
+    handle.close = (cb) => {
+      debug('close');
+      this.doClose(cb);
+    };
+    handle.isAlive = () => this.isAlive();
+    handle.isClosing = () => this.isClosing();
+    handle.onreadstart = () => this.readStart();
+    handle.onreadstop = () => this.readStop();
+    handle.onshutdown = (req) => this.doShutdown(req);
+    handle.onwrite = (req, bufs) => this.doWrite(req, bufs);
+
+    stream.pause();
+    stream.on('error', (err) => this.emit('error', err));
+    const ondata = (chunk) => {
+      if (typeof chunk === 'string' ||
+          stream._readableState.objectMode === true) {
+        // Make sure that no further `data` events will happen
+        stream.pause();
+        stream.removeListener('data', ondata);
+
+        this.emit('error', new errors.Error('ERR_STREAM_WRAP'));
+        return;
+      }
+
+      debug('data', chunk.length);
+      if (this._handle)
+        this._handle.readBuffer(chunk);
+    };
+    stream.on('data', ondata);
+    stream.once('end', () => {
+      debug('end');
+      if (this._handle)
+        this._handle.emitEOF();
+    });
+
+    super({ handle, manualStart: true });
+    this.stream = stream;
+    this._list = null;
+    this.read(0);
+  }
+
+  // Legacy
+  static get StreamWrap() {
+    return JSStreamWrap;
+  }
+
+  isAlive() {
+    return true;
+  }
+
+  isClosing() {
+    return !this.readable || !this.writable;
+  }
+
+  readStart() {
+    this.stream.resume();
+    return 0;
+  }
+
+  readStop() {
+    this.stream.pause();
+    return 0;
+  }
+
+  doShutdown(req) {
+    const handle = this._handle;
+    const item = this._enqueue('shutdown', req);
+
+    this.stream.end(() => {
+      // Ensure that write was dispatched
+      setImmediate(() => {
+        if (!this._dequeue(item))
+          return;
+
+        handle.finishShutdown(req, 0);
+      });
+    });
+    return 0;
+  }
+
+  doWrite(req, bufs) {
+    const self = this;
+    const handle = this._handle;
+
+    var pending = bufs.length;
+
+    // Queue the request to be able to cancel it
+    const item = this._enqueue('write', req);
+
+    this.stream.cork();
+    for (var n = 0; n < bufs.length; n++)
+      this.stream.write(bufs[n], done);
+    this.stream.uncork();
+
+    function done(err) {
+      if (!err && --pending !== 0)
+        return;
+
+      // Ensure that this is called once in case of error
+      pending = 0;
+
+      let errCode = 0;
+      if (err) {
+        const code = uv[`UV_${err.code}`];
+        errCode = (err.code && code) ? code : uv.UV_EPIPE;
+      }
+
+      // Ensure that write was dispatched
+      setImmediate(function() {
+        // Do not invoke callback twice
+        if (!self._dequeue(item))
+          return;
+
+        handle.doAfterWrite(req);
+        handle.finishWrite(req, errCode);
+      });
+    }
```

Comment:
can we avoid allocating two closures for every write? I know it would make the code uglier, but it's probably the right moment to do.

The innermost can be replaced by a top-level function with handle, item, req and errCode as parameters. The top-level one can be allocated in the constructor, and we can store `pending` as an object property (maybe behind a symbol).